contracts/TokenHolder.sol

```diff
@@ -43,465 +42,507 @@ contract TokenHolder is MultiSigWallet {
+        require(
+            status == AuthorizationStatus.AUTHORIZED &&
+            expirationHeight <= block.number,
+            "Key is not active."
+        );
+        _;
+    }
+
+    /** Requires that key is in authorized state. */
+    modifier keyIsAuthorized(address _key)
+    {
+        AuthorizationStatus status = ephemeralKeys[_key].status;
+        require(
+            status == AuthorizationStatus.AUTHORIZED,
+            "Key is not authorized."
+        );
+        _;
+    }
+
+    /**
+     * Requires that key was authorized. Key might be in authorized or
+     * revoked state.
+     */
+    modifier keyWasAuthorized(address _key)
+    {
+        AuthorizationStatus status = ephemeralKeys[_key].status;
+        require(
+            status != AuthorizationStatus.NOT_AUTHORIZED,
+            "Key was not authorized."
+        );
+        _;
+    }
+
+    /** Requires that key was not authorized. */
+    modifier keyWasNotAuthorized(address _key)
+    {
+        AuthorizationStatus status = ephemeralKeys[_key].status;
+        require(
+            status == AuthorizationStatus.NOT_AUTHORIZED,
+            "Key is not authorized."
+        );
+        _;
+    }
+
+    /** Requires that key has not expired. */
+    modifier keyHasNotExpired(address _key)
+    {
+        uint256 expirationHeight = ephemeralKeys[_key].expirationHeight;
+        require(
+            expirationHeight > block.number,
+            "Expiration key has expired."
+        );
+        _;
+    }
+
+
+    /* Special Functions */
+
     /**
-     * @notice Contract constructor.
-     *
      * @param _brandedToken erc20 contract address this user is part of.
      * @param _coGateway utility chain gateway contract address.
      * @param _tokenRules Token rules contract address.
      * @param _required No of requirements for multi sig wallet.
      * @param _wallets array of wallet addresses.
      */
-    constructor(
+    constructor (
         address _brandedToken,
         address _coGateway,
         address _tokenRules,
-        uint8 _required,
+        uint256 _required,
         address[] _wallets
     )
         public
         MultiSigWallet(_wallets, _required)
     {
         require(
             _brandedToken != address(0),
-            "Branded token contract address is 0"
+            "Branded token contract address is 0."
         );
         require(
             _coGateway != address(0),
-            "Co gateway contract address is 0"
+            "Co gateway contract address is 0."
         );
         require(
             _tokenRules != address(0),
-            "TokenRules contract address is 0"
+            "TokenRules contract address is 0."
         );
 
         brandedToken = _brandedToken;
         coGateway = _coGateway;
-        // Needed for onlyTokenRules contract validation
         tokenRules = _tokenRules;
     }
 
 
-    /* Public Functions */
+    /* External Functions */
```

Comment:
Below are public functions. Do we need to make it external? 
submitAuthorizeSession
submitRevokeSession .